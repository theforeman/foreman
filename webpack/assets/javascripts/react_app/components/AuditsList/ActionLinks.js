import React from 'react';
import { Col } from 'patternfly-react';
import PropTypes from 'prop-types';

const ActionLinks = ({ allowedActions }) => (
  <Col sm={2} className="actions-btns">
    {allowedActions &&
      allowedActions.map(
        ({ url, css_class: CssClassString, disabled, name, title }, index) => (
          <a
            key={index}
            {...{ className: CssClassString, href: url, disabled }}
          >
            {name || title}
          </a>
        )
      )}
  </Col>
);

ActionLinks.propTypes = {
  allowedActions: PropTypes.arrayOf(
    PropTypes.shape({
      url: PropTypes.string,
      title: PropTypes.string,
      name: PropTypes.string,
      css_class: PropTypes.string,
      disabled: PropTypes.bool,
    })
  ),
};

ActionLinks.defaultProps = {
  allowedActions: [],
};

export default ActionLinks;
