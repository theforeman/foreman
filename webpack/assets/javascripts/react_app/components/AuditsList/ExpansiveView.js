import React from 'react';
import PropTypes from 'prop-types';
import DiffContainer from '../DiffView/DiffContainer';
import { translate as __ } from '../../common/I18n';

const renderListItems = items =>
  items &&
  items.map((item, index) =>
    item && typeof item === 'string' && item.length > 0 ? (
      <tr key={index}>
        <td>{item}</td>
      </tr>
    ) : null
  );

const renderCols = changeArr =>
  changeArr &&
  changeArr.map(({ css_class: CssClassStr, id_to_label: idToLabel }, index) => (
    <td
      key={index}
      className={
        changeArr.length > 1
          ? `col-6 col-md-4 ${CssClassStr}`
          : `col-12 col-md-8 ${CssClassStr}`
      }
    >
      <div className={CssClassStr}>
        <p>{idToLabel}</p>
      </div>
    </td>
  ));

const renderTableRows = changeEntries =>
  changeEntries &&
  changeEntries.map(({ name, change }, index) => (
    <tr key={index}>
      <td key={index} className="col-6 col-md-4">
        <div>{name}</div>
      </td>
      {renderCols(change)}
    </tr>
  ));

const showAuditChanges = (
  actionDisplayName,
  auditedChangesWithIdToLabel,
  details
) => {
  const tableClasses = 'table table-bordered table-hover';

  if (['added', 'removed'].includes(actionDisplayName) && details.length > 0) {
    return (
      <table
        className={`${tableClasses} details-row table-inline-changes ${
          actionDisplayName === 'added' ? 'show-new' : 'show-old'
        }`}
      >
        <tbody>{renderListItems(details)}</tbody>
      </table>
    );
  }

  if (auditedChangesWithIdToLabel.length > 0) {
    return (
      <table className={`table-changes ${tableClasses}`}>
        <tbody>{renderTableRows(auditedChangesWithIdToLabel)}</tbody>
      </table>
    );
  }
  return null;
};

class ExpansiveView extends React.Component {
  showTemplateDiffIfAny() {
    const { template } = this.props.auditedChanges;
    if (template && template[0] !== template[1]) {
      return <DiffContainer oldText={template[0]} newText={template[1]} />;
    }
    return null;
  }

  render() {
    const {
      comment,
      actionDisplayName,
      auditedChangesWithIdToLabel,
      details,
    } = this.props;

    return (
      <div className="grid-container">
        {this.showTemplateDiffIfAny()}
        {showAuditChanges(
          actionDisplayName,
          auditedChangesWithIdToLabel,
          details
        )}
        {comment && (
          <div className="details-row comment-section">
            <p className="comment-title">
              <strong>{__('Comments')}</strong>
            </p>
            <p className="comment-desc">{comment}</p>
          </div>
        )}
      </div>
    );
  }
}

ExpansiveView.propTypes = {
  actionDisplayName: PropTypes.string.isRequired,
  auditedChanges: PropTypes.object.isRequired,
  comment: PropTypes.string,
  auditedChangesWithIdToLabel: PropTypes.arrayOf(
    PropTypes.shape({
      change: PropTypes.arrayOf(
        PropTypes.shape({
          css_class: PropTypes.string,
          id_to_label: PropTypes.string,
        })
      ),
      name: PropTypes.string,
    })
  ),
  details: PropTypes.arrayOf(PropTypes.string),
};

ExpansiveView.defaultProps = {
  comment: undefined,
  auditedChangesWithIdToLabel: [],
  details: undefined,
};

export default ExpansiveView;
