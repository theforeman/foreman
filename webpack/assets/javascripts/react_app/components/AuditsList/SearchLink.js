import React from 'react';
import PropTypes from 'prop-types';
import EllipsisWithTooltip from 'react-ellipsis-with-tooltip';

const SearchLink = ({ url, title, id, textValue }) => {
  const linkProps = {
    href: url,
    title,
    id: `resource-link-${id}`,
  };

  return (
    <EllipsisWithTooltip>
      <a {...linkProps}>{textValue}</a>
    </EllipsisWithTooltip>
  );
};

SearchLink.propTypes = {
  url: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  title: PropTypes.string,
  textValue: PropTypes.string,
};

SearchLink.defaultProps = {
  title: undefined,
  textValue: '',
};

export default SearchLink;
