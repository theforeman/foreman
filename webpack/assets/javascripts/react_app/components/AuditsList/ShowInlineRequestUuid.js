import React from 'react';
import PropTypes from 'prop-types';
import { Row, Col } from 'patternfly-react';
import EllipsisWithTooltip from 'react-ellipsis-with-tooltip';
import { translate as __ } from '../../common/I18n';

const ShowInlineRequestUuid = ({ fetchAndPush, requestUuid, id }) => (
  <Col sm={10} className="request-uuid-column">
    <Row>
      <Col md={2}>
        <span>{__('Request UUID')}</span>
      </Col>
      <Col md={10} className="value">
        <EllipsisWithTooltip>
          <a
            onClick={() =>
              fetchAndPush({ searchQuery: `request_uuid = ${requestUuid}` })
            }
            title={__(
              'HTTP request UUID, clicking will filter audits for this request. It can also be used for searching in application logs.'
            )}
          >
            {requestUuid}
          </a>
        </EllipsisWithTooltip>
      </Col>
    </Row>
  </Col>
);

ShowInlineRequestUuid.propTypes = {
  fetchAndPush: PropTypes.func.isRequired,
  requestUuid: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
};

export default ShowInlineRequestUuid;
