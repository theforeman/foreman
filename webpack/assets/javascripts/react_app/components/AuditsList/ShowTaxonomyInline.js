import React from 'react';
import PropTypes from 'prop-types';
import { Row, Col } from 'patternfly-react';

const ShowTaxonomyInline = ({ displayLabel, items }) => {
  const listItems = items.map(
    ({ name, url, disabled, css_class: addCSS }, index) => (
      <a
        href={url}
        key={index}
        className={`apply-comma ${addCSS || ''}`}
        disabled={disabled}
      >
        {name}
      </a>
    )
  );

  return (
    <Row>
      <Col md={2}>
        <span>{displayLabel}</span>
      </Col>
      <Col md={10}>
        <strong>{items && listItems}</strong>
      </Col>
    </Row>
  );
};

ShowTaxonomyInline.propTypes = {
  displayLabel: PropTypes.string,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      url: PropTypes.string,
      css_class: PropTypes.string,
      disabled: PropTypes.bool,
    })
  ),
};

ShowTaxonomyInline.defaultProps = {
  displayLabel: '',
  items: [],
};

export default ShowTaxonomyInline;
