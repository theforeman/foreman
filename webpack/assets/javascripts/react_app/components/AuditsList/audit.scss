#audit-list {
  .apply-comma:not(:last-child)::after {
    white-space: pre;
    content: ", ";
  }

  .list-view-pf-view {
    margin-top: 10px;
  }

  .list-view-pf-body > .list-view-pf-description {
    width: 30%;
    flex: 1 0 30%;
  }

  .list-group-item.list-view-pf-expand-active {
    background-color: #def3ff;
  }

  .list-group-item {
    &:hover {
      background-color: #edf8ff;
    }

    .list-group-item-header {
      .list-group-item-heading {
        flex: 1 0 unquote("calc(30% - 20px)");
        width: unquote("calc(30% - 20px)");

        .user-grid {
          display: grid;

          .user-info {
            font-weight: bold;
          }
        }
      }

      .list-view-pf-additional-info-item {
        font-weight: 600;
        flex: 2;
        justify-content: flex-end;
        text-align: right;
        max-width: inherit;

        &:last-child {
          justify-content: flex-start;
          text-align: left;
          flex-grow: 3;
          overflow: hidden;
          overflow-wrap: break-word;
          text-overflow: ellipsis;
          white-space: nowrap;
          word-break: break-all;
        }
      }

      .list-view-pf-actions {
        margin-bottom: 0;
        margin-top: 22px;
      }

      .list-view-pf-expand {
        margin-right: 35px;
        margin-bottom: 10px;
        margin-top: 18px;
      }
    }
  }

  .list-group-item-text {
    width: calc(65% - 40px);

    .list-view-pf-description {
      .list-group-item-text {
        padding-left: 30px;
      }
    }
  }

  .main-info-minimize-padding > .list-group-item-header {
    .list-view-pf-main-info {
      padding-bottom: 10px;
      padding-top: 12px;
    }
  }

  .main-info-maximize-padding > .list-group-item-header {
    .list-view-pf-main-info {
      padding-bottom: 21px;
      padding-top: 21px;
    }
  }

  .gray-text {
    color: #999;
  }

  .list-group-item-container {
    padding-top: 20px;
    padding-left: 35px;

    .actions-btns {
      padding-bottom: 15px;

      .btn {
        float: right;
        margin-left: 10px;
      }
    }

    .taxonomy-section,
    .request-uuid-column {
      padding-left: 50px;
      word-break: break-word;
    }

    .request-uuid-column .value {
      font-weight: bold;
    }

    .grid-container {
      .details-row {
        border: 1px solid #d1d1d1;
        margin-top: 15px;
      }

      .comment-section {
        p {
          margin-left: 10px;
          margin-right: 15px;
        }

        .comment-title {
          margin-top: 10px;
        }
      }

      .table-inline-changes {
        margin-top: 10px;
      }

      .table-changes {
        margin-top: 15px;

        td {
          p {
            margin: 0 15px;
          }
        }
      }

      .table-changes,
      .table-inline-changes {
        word-break: break-word;
        font-weight: 600;

        td {
          padding: 5px;
        }

        td:first-child {
          padding-left: 30px;
        }
      }

      .table-hover > tbody > tr:hover,
      .table-hover > tbody > tr:hover > td {
        border: 1px solid #d1d1d1;
      }

      .table-hover > tbody > tr:hover,
      .table-hover > tbody > tr:hover > td:first-child {
        background-color: #fff;
      }

      .show-old {
        background-color: #ffeef0;
      }

      .show-new {
        background-color: #e6ffed;
      }

      .table-inline-changes.table-hover.show-new > tbody > tr:hover > td,
      .table-inline-changes.table-hover.show-new > tbody > tr:hover > th {
        background-color: #e6ffed;
      }

      .table-inline-changes.table-hover.show-old > tbody > tr:hover > td,
      .table-inline-changes.table-hover.show-old > tbody > tr:hover > th {
        background-color: #ffeef0;
      }

      .editor-section {
        margin-top: 15px;

        .editor.ace_editor {
          border: 1px solid #d1d1d1;
        }
      }
    }
  }
}
