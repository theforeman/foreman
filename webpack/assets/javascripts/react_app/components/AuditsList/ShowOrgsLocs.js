import React from 'react';
import PropTypes from 'prop-types';
import { Col } from 'patternfly-react';
import ShowTaxonomyInline from './ShowTaxonomyInline';
import { translate as __ } from '../../common/I18n';

const ShowOrgsLocs = ({ orgs, locs }) => (
  <Col sm={10} className="taxonomy-section">
    <ShowTaxonomyInline
      displayLabel={__('Affected Organizations')}
      items={orgs}
    />
    <ShowTaxonomyInline displayLabel={__('Affected Locations')} items={locs} />
  </Col>
);

ShowOrgsLocs.propTypes = {
  orgs: PropTypes.array,
  locs: PropTypes.array,
};

ShowOrgsLocs.defaultProps = {
  orgs: [],
  locs: [],
};

export default ShowOrgsLocs;
