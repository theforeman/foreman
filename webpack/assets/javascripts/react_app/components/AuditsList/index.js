import React from 'react';
import PropTypes from 'prop-types';
import { ListView, Row } from 'patternfly-react';
import SearchLink from './SearchLink';
import ShowInlineRequestUuid from './ShowInlineRequestUuid';
import ShowOrgsLocs from './ShowOrgsLocs';
import ActionLinks from './ActionLinks';
import ExpansiveView from './ExpansiveView';
import UserDetails from './UserDetails';
import { translate as __ } from '../../common/I18n';
import ShortDateTime from '../common/dates/ShortDateTime';
import './audit.scss';

const isAuditLogin = auditedChanges => {
  let name;
  try {
    [name] = Object.keys(auditedChanges);
  } catch (e) {
    name = '';
  }
  return name === 'last_login_on';
};

const description = actionDisplayName => (
  <ListView.Description>
    <ListView.DescriptionText>{actionDisplayName}</ListView.DescriptionText>
  </ListView.Description>
);

const renderAdditionalInfoItems = items =>
  items &&
  items.map((item, index) => (
    <ListView.InfoItem key={index}>{item}</ListView.InfoItem>
  ));

const renderTimestamp = date => (
  <span className="gray-text">
    <ShortDateTime
      date={date}
      defaultValue={__('N/A')}
      showRelativeTimeTooltip
    />
  </span>
);

const renderResourceLink = (auditTitle, auditTitleUrl, id) => {
  if (auditTitleUrl) {
    return (
      <SearchLink
        url={auditTitleUrl}
        textValue={auditTitle}
        title={__('Filter audits for this resource only')}
        id={id}
      />
    );
  }
  return auditTitle;
};

const AuditsList = ({ data: { audits }, fetchAndPush }) => {
  const initExpanded = audits.length === 1;
  return (
    <ListView>
      {audits.map(
        ({
          id,
          created_at: createdAt,
          audited_type_name: auditedTypeName,
          audit_title: auditTitle,
          audit_title_url: auditTitleUrl,
          audited_changes: auditedChanges,
          user_info: userInfo,
          remote_address: remoteAddress,
          action_display_name: actionDisplayName,
          affected_organizations: affectedOrganizations,
          affected_locations: affectedLocations,
          allowed_actions: allowedActions,
          request_uuid: requestUuid,
          comment,
          audited_changes_with_id_to_label: auditedChangesWithIdToLabel,
          details,
        }) => (
          <ListView.Item
            id={id}
            key={id}
            className={
              remoteAddress
                ? 'main-info-minimize-padding'
                : 'main-info-maximize-padding'
            }
            actions={renderTimestamp(createdAt)}
            additionalInfo={renderAdditionalInfoItems([
              auditedTypeName.toUpperCase(),
              renderResourceLink(auditTitle, auditTitleUrl, id),
            ])}
            heading={
              <UserDetails
                isAuditLogin={isAuditLogin(auditedChanges)}
                userInfo={userInfo}
                remoteAddress={remoteAddress}
              />
            }
            description={description(actionDisplayName)}
            stacked={false}
            hideCloseIcon
            initExpanded={initExpanded}
          >
            <Row>
              <ShowOrgsLocs
                orgs={affectedOrganizations}
                locs={affectedLocations}
              />
              <ActionLinks allowedActions={allowedActions} />
            </Row>

            <Row>
              <ShowInlineRequestUuid
                fetchAndPush={fetchAndPush}
                requestUuid={requestUuid}
                id={id}
              />
            </Row>

            <ExpansiveView
              {...{
                actionDisplayName,
                details,
                comment,
                auditTitle,
                auditedChanges,
                auditedChangesWithIdToLabel,
              }}
            />
          </ListView.Item>
        )
      )}
    </ListView>
  );
};
AuditsList.propTypes = {
  data: PropTypes.shape({
    audits: PropTypes.array.isRequired,
  }).isRequired,
  fetchAndPush: PropTypes.func.isRequired,
};

export default AuditsList;
