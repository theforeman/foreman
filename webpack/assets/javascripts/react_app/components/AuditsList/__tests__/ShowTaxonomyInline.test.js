import toJson from 'enzyme-to-json';
import { shallowRenderComponentWithFixtures } from '../../../common/testHelpers';
import ShowTaxonomyInline from '../ShowTaxonomyInline';
import { TaxonomyProps } from './AuditsList.fixtures';
import { translate as __ } from '../../../common/I18n';

const OrgsFixtures = {
  'render organizations': {
    displayLabel: __('Affected Organizations'),
    items: TaxonomyProps.orgs,
  },
};

describe('ShowTaxonomyInline', () => {
  describe('rendering', () => {
    const components = shallowRenderComponentWithFixtures(
      ShowTaxonomyInline,
      OrgsFixtures
    );
    components.forEach(({ description, component }) => {
      it(description, () => {
        expect(toJson(component)).toMatchSnapshot();
      });
    });
  });
});
