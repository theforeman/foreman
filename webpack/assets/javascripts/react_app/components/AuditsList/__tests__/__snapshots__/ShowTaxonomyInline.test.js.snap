// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowTaxonomyInline rendering render organizations 1`] = `
<Row
  bsClass="row"
  componentClass="div"
>
  <Col
    bsClass="col"
    componentClass="div"
    md={2}
  >
    <span>
      Affected Organizations
    </span>
  </Col>
  <Col
    bsClass="col"
    componentClass="div"
    md={10}
  >
    <strong>
      <a
        className="apply-comma "
        href="/organizations/1-testOrg/edit"
        key="0"
      >
        testOrg
      </a>
      <a
        className="apply-comma disabled"
        disabled={true}
        href="#"
        key="1"
      >
        testOrg2
      </a>
    </strong>
  </Col>
</Row>
`;
